### ISSUE TITLE
lint spurious searches

### ISSUE THREAD

**Issue by llogiq** [Manishearth/rust-clippy#865]
iter loops that set a variable to the current element based on an equality check, but do not `break`, suggest `.iter().find(..)` or `.iter().rev().find(..)`, e.g.

```
let mut found = None;
for x in xs {
    if x == f {
        found = Some(x);
    }
}
```

**Comment by devonhollowood** [comment]
Could this change semantics if you have two different values which compare as equal for some reason? For example:

```rust
#[derive(Debug)]
struct CaseInsensitiveChar(char);

impl PartialEq<char> for CaseInsensitiveChar {
    fn eq(&self, other: &char) -> bool {
        let CaseInsensitiveChar(a) = *self;
        a.to_lowercase().eq(other.to_lowercase())
    }
}

fn main() {
    let mut last_a = CaseInsensitiveChar('?');
    for c in "Abracadabra".chars().map(|x| CaseInsensitiveChar(x)) {
        if c == 'A' {
            last_a = c;
            // inserting a break here changes the result from an 'a' to an 'A'.
        }
    }
    println!("{:?}", last_a);
}
```

I could see the counter argument that in this case you are treating 'A' and 'a' as equivalent, so it shouldn't matter which you get out of the loop. And I suppose that if one really wants to allow this edge-case behavior, one could always add `#[allow(spurious_search)]` or whatever. But I think this case is interesting enough to bring up for discussion.

**Comment by llogiq** [comment]
This would still be equal to `"Abracadabra".chars().rev().find(..)`. And if you don't care about direction, you can omit the reversal.

**Comment by felix91gr** [comment]
> Should this be a lint? Many times when writing a manual find like this the lack of a `break` is intentional for performance reasons.

@GnomedDev wait, how does that happen?

If I understand how the compiler treats iterator operations, they tend to end up equally or better optimized than naked loops, because the compiler ends up having access to more information about the semantics of the code.

There's probably a counterexample I'm not seeing. Please illuminate me :3

---

Though then again, lints can always be turned off (or on) depending on the user's preference.

**Comment by felix91gr** [comment]
> It can be faster if it removes a branch from the loop body.
> 
> for &x in items {
>   if x == key {
>     found = Some(x);
>   }
> }
> 
> This can be compiled into a cmov in the loop body or even turn into a vectorised search. Both will perform better in some cases.

Sorry to ask you this. I've been trying to conjure up a Playground to compare an iterator implementation of this side by side this example, but LLVM keeps just erasing the function in the generated code. I'm guessing this is doomed to happen due to playgrounds functioning with mostly constant values.

Do you know how I could make two such implementations (one like the loop you show and one with an iterator's `find`), such that you, me and whomever passes by this issue in the future can all see the same compilation results?

I want to do this because optimizing compilers are really good at what they do, and iterator operations are perfect for their static analysis. Therefore I want to prove that this example gets optimized better than its equivalent in the iterator world.

**Comment by Jarcho** [comment]
https://godbolt.org/ is normally what I would for asm output. Generally better to use that then the playground.

**Comment by felix91gr** [comment]
> https://godbolt.org/ is normally what I would for asm output. Generally better to use that than the playground.

Yes, but if we're measuring asm performance post-LLVM-optimizations, then a simple script in godbolt will get const-eval'd into a constant in the print statement where we use it. How would you test the performance (**post optimization!**) of these two scripts?

```rust
let mut found = None;
for &x in items {
  if x == key {
    found = Some(x);
  }
}
```

and

```rust
let found = items.find(key);
```

Because for both cases, let's say that:

```rust
items = vec![5, 4, 3, 2, 1];
key = 2;

// Code pertaining to finding the key

let found = found.unwrap(); // Yes I know I'm unwrapping, but this is so that the printed value isn't an Option value

println!("This is what I've got: {found}");
```

If we compiled all of that in Release mode, with optimizations enabled, then the resulting code will be the equivalent of the ASM of this:

```rust
println!("This is what I've got: 2");
```

Regarless of whether we found the key with the loop or with the iterator functions.

That's what I mean @Jarcho. What method would you recommend I use for comparing these without the assembly being const-eval'd into oblivion?

**Comment by Jarcho** [comment]
Assuming you just want to see the asm you can take things as function arguments. e.g. https://godbolt.org/z/1EvhbEe7W

**Comment by felix91gr** [comment]
Anywho, assuming this allows us to compare apples to apples, this is what I've got: https://godbolt.org/z/MPEfMzh9e

That's what (up to what I understand) @llogiq was arguing for in the original example.

**Comment by Jarcho** [comment]
The `no_mangle` attribute prevents it from being removed.
